modules: throw an error for invalid package.json main entries

BridgeAR opened the pull request:
We currently ignore invalid `main` entries in package.json files.
This does not seem to be very user friendly as it's certainly an
error if the `main` entry is not a valid file name. So instead of
trying to resolve the file otherwise, throw an error immediately to
improve the user experience.

Fixes: https://github.com/nodejs/node/issues/26588

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/21691/
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1770/

guybedford commented on doc/api/errors.md:
```diff
@@ -1463,6 +1463,12 @@ The V8 platform used by this instance of Node.js does not support creating
 Workers. This is caused by lack of embedder support for Workers. In particular,
 this error will not occur with standard builds of Node.js.
 
+<a id="ERR_MODULE_NOT_FOUND"></a>
```
Resolver errors always have code "MODULE_NOT_FOUND". Changing this to "ERR_MODULE_NOT_FOUND" just for one specific case seems both inconsistent and breaking to me.

For example, lots of code has guards that do `try { require.resolve(x) } catch (e) { if (e.code === 'MODULE_NOT_FOUND' }` sort of thing.

BridgeAR replied:
I wondered why we did not yet change that error code but that makes sense. Is this something we ever plan to fix (as in: align with all other error codes in Node.js)?

BridgeAR replied:
I changed it to `MODULE_NOT_FOUND`.

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -225,15 +226,26 @@ function readPackage(requestPath) {
+  const actual = tryFile(filename, isMain) ||
+    tryExtensions(filename, exts, isMain) ||
+    tryExtensions(path.resolve(filename, 'index'), exts, isMain);
+  if (actual === false) {
```
I haven't read this closely enough, but does this also catch cases like `pkg/x.js` as well as mains?

BridgeAR replied:
It catches all main entries that we can not resolve.

vsemozhetbyt commented on doc/api/errors.md:
```diff
@@ -1463,6 +1463,12 @@ The V8 platform used by this instance of Node.js does not support creating
+<a id="ERR_MODULE_NOT_FOUND"></a>
+### ERR_MODULE_NOT_FOUND
+
+A failure occurred resolving a require call. This happens for example with a
```
```suggestion
A failure occurred resolving a require call. This happens for example with an
```

BridgeAR commented:
@guybedford that observation is correct. The code would theoretically run but it does seem like the expectation is that the other module would be loaded instead. So it does seem to be a clear user mistake. If someone provides an invalid `main` entry, we should notify the user about the mistake in what ever code that happens in.

CITGM actually revealed one failure: https://github.com/dominictarr/assertions seems to have an invalid `main` entry but it has an `index.js` in the directory which is resolved instead (I am going to open a PR to fix that).

I could add a runtime deprecation for cases that resolve with the `index.js` file being in the same directory (that's why I marked it as WIP, I wanted to give it some more thought) or should we throw the error right away? I guess having a transition time would be good. I do not fear files being able to resolve in other `node_modules` directories and more about this case.

@nodejs/modules PTAL.

jdalton commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -225,15 +225,27 @@ function readPackage(requestPath) {
+    err.path = path.resolve(requestPath, 'package.json');
+    err.requestPath = originalPath;
+    // TODO(BridgeAR): Add the requireStack as well.
+    throw err;
```
The test checks for `code: 'MODULE_NOT_FOUND'` but I don't see it set 👆

BridgeAR commented:
@guybedford I'll change it to a deprecation warning in case the main entry is invalid but there is a index.js file in the same directory as well.

guybedford commented:
That seems sensible to me.

guybedford commented:
One thing that could be quite frustrating is, as per usual, if the warning is in a dependency the user doesn't necessarily have control over.

Perhaps we should docs-only deprecate first as per policy?

Which makes me wish there could be a more gradual way to introduce warnings... say a `node --latest-warnings` or similar, where only those users interested in being ahead can do the checks, then get the fixes published to the third-party packages for everyone else before they become annoyance.

targos commented:
> Which makes me wish there could be a more gradual way to introduce warnings...

We have the `--pending-deprecation` flag.

BridgeAR commented:
I updated the PR to keep on working in case the `index.js` file is in the same directory as the package.json.

Now, only completely broken modules throw the error (as in: no file could be found what so ever). On top of that I added a pending deprecation in case the `main` entry is invalid.

richardlau commented:
This changes the high-level algorithm described in https://github.com/nodejs/node/blob/master/doc/api/modules.md#all-together. Please could you update that to match?

BridgeAR commented:
@richardlau we did not document potential parsing errors so far either. Shall I include that as well?

vsemozhetbyt commented on doc/api/deprecations.md:
```diff
@@ -2368,6 +2368,22 @@ Type: Runtime
 The `_stream_wrap` module is deprecated.
 
+<a id="DEP0XXX"></a>
+### DEP0XXX: modules with an invalid `main` entry and a `index.js` file
```
```suggestion
### DEP0XXX: modules with an invalid `main` entry and an `index.js` file
```

vsemozhetbyt commented on doc/api/deprecations.md:
```diff
@@ -2368,6 +2368,22 @@ Type: Runtime
+
+Type: Runtime (supports [`--pending-deprecation`][])
+
+Modules that have an invalid `main` entry (e.g., `./does-not-exist.js`) that
```
Maybe:
```suggestion
Modules that have an invalid `main` entry (e.g., `./does-not-exist.js`) and
```

vsemozhetbyt commented on doc/api/deprecations.md:
```diff
@@ -2368,6 +2368,22 @@ Type: Runtime
+Type: Runtime (supports [`--pending-deprecation`][])
+
+Modules that have an invalid `main` entry (e.g., `./does-not-exist.js`) that
+also have a `index.js` file in the top level directory will resolve the
```
```suggestion
also have an `index.js` file in the top level directory will resolve the
```

BridgeAR commented:
PTAL. I updated the documentation accordingly. It now includes the steps that we took but that were never documented so far.

tniessen commented on doc/api/deprecations.md:
```diff
@@ -2368,6 +2368,22 @@ Type: Runtime
+    description: Runtime deprecation.
+-->
+
+Type: Runtime (supports [`--pending-deprecation`][])
```
This is not what we usually do. The type should be documentation-only as long as the default behavior does not match "Runtime". The next step would be to upgrade from `Documentation-only (supports --pending-deprecation)` to `Runtime` which should behave as if node was started with `--pending-deprecation`.

BridgeAR replied:
Thanks, I had a look at the buffer constructor one and forgot that it's a special case. I updated it to `Documentation-only`.

tniessen commented on doc/api/deprecations.md:
```diff
@@ -2368,6 +2368,22 @@ Type: Runtime
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/26823
+    description: Runtime deprecation.
```
Same here, sorry, I should have mentioned that.

jdalton commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -225,15 +226,41 @@ function readPackage(requestPath) {
+      const packageJSON = path.resolve(requestPath, 'package.json');
+      process.emitWarning(
+        `The 'package.json' (${packageJSON}) has an invalid 'main' entry ` +
+          `(${pkg}). Please either fix that or report it to the module author.`,
```
`entry` -> `field`.

I think writing `'package.json'` and then putting a path to the `package.json` is a bit of duplication. Maybe it could be worded like:

```js
`Invalid 'main' field in '${jsonPath}' of '${fieldValue}'. ` +
`Please either fix that or report it to the module author`
```

Also note no period on the second sentence.

BridgeAR commented:
Rebased due to conflicts.

CI https://ci.nodejs.org/job/node-test-pull-request/21843/
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1773/

Sadly using https://github.com/nodejs/Gzemnid won't work here currently. @ChALkeR I did not look into the logic so far but it should actually be relatively straight forward to check against invalid `main` fields in `package.json` while doing a generic lookup. Would you mind having a look at that?

@nodejs/tsc PTAL, this is a semver-major change that already got some LGs but it still requires a sign-off from the TSC.

Trott commented:
> I wouldn't mind if you added this to `--pending-deprecations` so it would flush out examples of this for people (like me) who run with pending deprecations enabled.

Er... I mean, "thanks for doing that"...didn't update the text after I finished reviewing. :-D

BridgeAR commented:
Resumed CI for a yellow build: https://ci.nodejs.org/job/node-test-pull-request/21870/

CITGM looks good to me.

BridgeAR commented on doc/api/deprecations.md:
```diff
@@ -2400,6 +2400,22 @@ Please use the publicly documented [`timeout.refresh()`][] instead.
+Modules that have an invalid `main` entry (e.g., `./does-not-exist.js`) and
+also have an `index.js` file in the top level directory will resolve the
+`index.js` file. That is deprecated and is going to throw an error in future
+Node.js versions.
```
Only in case the `--pending-deprecation` flag is used.

BridgeAR commented:
Thanks a lot for the reviews.

Landed in 115f0f5a57f50f6b039f28a56910207f92df116d :tada:
